test/integration/targets/mongodb_shard/tasks/main.yml

```diff
@@ -0,0 +1,454 @@
+  wait_for:
+    host: localhost
+    port: 4000
+    delay: 1
+
+- name: Add shard 1
+  mongodb_shard:
+    login_user: "{{ mongodb_admin_user }}"
+    login_password: "{{ mongodb_admin_password }}"
+    shard: "{{ mongodb_replicaset1 }}/localhost:3001"
+    state: present
+
+- name: Add shard 2
+  mongodb_shard:
+    login_user: "{{ mongodb_admin_user }}"
+    login_password: "{{ mongodb_admin_password }}"
+    shard: "{{ mongodb_replicaset2 }}/localhost:3004"
+    state: present
+
+- name: Test with bad password
+  mongodb_shard:
+    login_user: "{{ mongodb_admin_user }}"
+    login_password: XXXXXXXXXXXX
+    shard: "{{ mongodb_replicaset2 }}/localhost:3004"
+    state: present
+  register: mongodb_shard_bad_pw
+  ignore_errors: True
+
+- name: Assert login failed
+  assert:
+    that:
+      - "mongodb_shard_bad_pw.changed == False"
+      - "'unable to connect to database: Authentication failed.' == mongodb_shard_bad_pw.msg"
+
+- name: Get replicaset info
+  command: mongo admin --eval "sh.status()" --port 27017 -u "{{ mongodb_admin_user }}" -p "{{ mongodb_admin_password }}"
+  register: mongo_output
+
+- name: Assert shard name is in mongo_output
+  assert:
+    that:
+      - "mongo_output.changed == true"
+      - "'{{ mongodb_replicaset1 }}/localhost:3001,localhost:3002,localhost:3003' in mongo_output.stdout"
+      - "'{{ mongodb_replicaset2 }}/localhost:3004,localhost:3005,localhost:3006' in mongo_output.stdout"
+      - "'balancer' in mongo_output.stdout"
+
+- name: Remove shard 2
+  mongodb_shard:
+    login_user: "{{ mongodb_admin_user }}"
+    login_password: "{{ mongodb_admin_password }}"
+    shard: "{{ mongodb_replicaset2 }}"
+    state: absent
+
+- name: Get replicaset info
+  command: mongo admin --eval "sh.status()" --port 27017 -u "{{ mongodb_admin_user }}" -p "{{ mongodb_admin_password }}"
+  register: mongo_output
+
+- name: Assert shard 2 is draining
+  assert:
+    that:
+      - "mongo_output.changed == true"
+      - "'{{ mongodb_replicaset1 }}/localhost:3001,localhost:3002,localhost:3003' in mongo_output.stdout"
+      - "'\"draining\" : true' in mongo_output.stdout"
+      - "'balancer' in mongo_output.stdout"
+
+- name: Run remove command again to finalize shard removal
+  mongodb_shard:
+    login_user: "{{ mongodb_admin_user }}"
+    login_password: "{{ mongodb_admin_password }}"
+    shard: "{{ mongodb_replicaset2 }}"
+    state: absent
+
+- name: Get replicaset info
+  command: mongo admin --eval "sh.status()" --port 27017 -u "{{ mongodb_admin_user }}" -p "{{ mongodb_admin_password }}"
+  register: mongo_output
+
+- name: Assert shard 2 is not present
+  assert:
+    that:
+      - "mongo_output.changed == true"
+      - "'{{ mongodb_replicaset1 }}/localhost:3001,localhost:3002,localhost:3003' in mongo_output.stdout"
+      - "'{{ mongodb_replicaset2 }}/localhost:3004,localhost:3005,localhost:3006' not in mongo_output.stdout"
+      - "'balancer' in mongo_output.stdout"
+
+- name: Run test with unknown host
```

Comment:
Same as above, can you set a lower `serverSelectionTimeoutMS` on the pymongo client, at least for this task? The default of 30 seconds adds a significant delay to the tests.

Reply:
I've excluded these tests for now. This will require changes to the module itself. I'd rather do that in another PR.